win32/devdoc/file_win32_requirements.md

```diff
@@ -0,0 +1,166 @@
+# file_win32
+================
+
+## Overview
+
+Windows implementation of the `file` module.
+
+## Exposed API
+
+```c
+#define FILE_WRITE_ASYNC_VALUES \
+    FILE_WRITE_ASYNC_INVALID_ARGS, \
+    FILE_WRITE_ASYNC_WRITE_ERROR, \
+    FILE_WRITE_ASYNC_ERROR,\
+    FILE_WRITE_ASYNC_OK
+MU_DEFINE_ENUM(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_VALUES);
+
+#define FILE_READ_ASYNC_VALUES \
+    FILE_READ_ASYNC_INVALID_ARGS, \
+    FILE_READ_ASYNC_READ_ERROR, \
+    FILE_READ_ASYNC_ERROR,\
+    FILE_READ_ASYNC_OK
+MU_DEFINE_ENUM(FILE_READ_ASYNC_RESULT, FILE_READ_ASYNC_VALUES);
+
+typedef struct FILE_HANDLE_DATA_TAG* FILE_HANDLE;
+typedef void(*FILE_REPORT_FAULT)(void* user_report_fault_context, const char* information);
+
+typedef void(*FILE_WRITE_CB)(void* user_context, bool is_successful);
+typedef void(*FILE_READ_CB)(void* user_context, bool is_successful, CONSTBUFFER_HANDLE content);
+MOCKABLE_FUNCTION(, FILE_HANDLE, file_create, EXECUTION_ENGINE_HANDLE, execution_engine, const char*, full_file_name, FILE_REPORT_FAULT, user_report_fault_callback, void*, user_report_fault_context);
+MOCKABLE_FUNCTION(, void, file_destroy, FILE_HANDLE, handle);
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_WRITE_ASYNC_RESULT, file_write_async, FILE_HANDLE, handle, CONSTBUFFER_HANDLE, source, uint64_t, position, FILE_WRITE_CB, user_callback, void*, user_context)(0, MU_FAILURE);
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_READ_ASYNC_RESULT, file_read_async, FILE_HANDLE, handle, uint32_t, size, uint64_t, position, FILE_READ_CB, user_callback, void*, user_context)(0, MU_FAILURE);
+MOCKABLE_FUNCTION_WITH_RETURNS(, int, file_extend_filesize, FILE_HANDLE, handle, uint64_t, desired_size, bool, has_manage_volume)(0, MU_FAILURE);
+```
+
+## file_create
+
+```c
+MOCKABLE_FUNCTION(, FILE_HANDLE, file_create, EXECUTION_ENGINE_HANDLE, execution_engine, const char*, full_file_name, FILE_REPORT_FAULT, user_report_fault_callback, void*, user_report_fault_context);
+```
+
+**SRS_FILE_WIN32_43_040: [** If `execution_engine` is `NULL`, `file_create` shall fail and return `NULL`. **]**
+
+**SRS_FILE_WIN32_43_041: [** If `full_file_name` is `NULL` then `file_create` shall fail and return `NULL`. **]**
+
+**SRS_FILE_WIN32_43_001: [** `file_create` shall call `CreateFileA` with `full_file_name` as `lpFileName`, `GENERIC_READ|GENERIC_WRITE` as `dwDesiredAccess`, `FILE_SHARED_READ` as `dwShareMode`, `NULL` as `lpSecurityAttributes`, `OPEN_ALWAYS` as `dwCreationDisposition`, `FILE_FLAG_OVERLAPPED|FILE_FLAG_WRITE_THROUGH` as `dwFlagsAndAttributes` and `NULL` as `hTemplateFiles`. **]**
+
+**SRS_FILE_WIN32_43_041: [** `file_create` shall allocate a `FILE_HANDLE`. **]**
+
+**SRS_FILE_WIN32_43_002: [** `file_create` shall call `SetFileCompletionNotificationModes` to disable calling the completion port when an async operations finishes synchrounously. **]**
+
+**SRS_FILE_WIN32_43_003: [** `file_create` shall initialize a threadpool environment. **]**

```

Comment:
>initialize [](start = 49, length = 10)

Is this threadpool environment somewhere on the stack... or is it allocated?

This may be important, because later it is =linked by registration with the threadpool and I am not usre if the environment goes away if the threadpool will be happy (likely not)

Reply:
Yeah this should be allocated. Changing in next iteration.

---
In reply to: [450485165](https://github.com/Azure/azure-c-pal/pull/7#discussion_r450485165) [](ancestors = 450485165)